pkg/cri/server/instrumented_service.go

```diff
@@ -455,6 +1497,48 @@ func (in *instrumentedService) Version(ctx context.Context, r *runtime.VersionRe
 	return res, errdefs.ToGRPC(err)
 }
 
+func (in *instrumentedAlphaService) Version(ctx context.Context, r *runtime_alpha.VersionRequest) (res *runtime_alpha.VersionResponse, err error) {
```

Comment:
Try to use function to handle alpha <-> v1 like

```diff
diff --git a/pkg/cri/server/instrumented_service.go b/pkg/cri/server/instrumented_service.go
index ee88277ee..6ed05dc57 100644
--- a/pkg/cri/server/instrumented_service.go
+++ b/pkg/cri/server/instrumented_service.go
@@ -128,6 +128,36 @@ func (in *instrumentedAlphaService) RunPodSandbox(ctx context.Context, r *runtim
 	return res, errdefs.ToGRPC(err)
 }

+func alphaReqToV1Req(
+	alphar interface{ Marshal() ([]byte, error) },
+	v1r interface{ Unmarshal(_ []byte) error },
+) error {
+	p, err := alphar.Marshal()
+	if err != nil {
+		return err
+	}
+
+	if err = v1r.Unmarshal(p); err != nil {
+		return err
+	}
+	return nil
+}
+
+func v1RespToAlphaResp(
+	v1res interface{ Marshal() ([]byte, error) },
+	alphares interface{ Unmarshal(_ []byte) error },
+) error {
+	p, err := v1res.Marshal()
+	if err != nil {
+		return err
+	}
+
+	if err = alphares.Unmarshal(p); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (in *instrumentedService) ListPodSandbox(ctx context.Context, r *runtime.ListPodSandboxRequest) (res *runtime.ListPodSandboxResponse, err error) {
 	if err := in.checkInitialized(); err != nil {
 		return nil, err
@@ -1512,33 +1542,22 @@ func (in *instrumentedAlphaService) Version(ctx context.Context, r *runtime_alph
 		}
 	}()
 	// converts request and response for earlier CRI version to call and get response from the current version
-	p, err := r.Marshal()
-	if err == nil {
-		var v1r runtime.VersionRequest
-		if err = v1r.Unmarshal(p); err == nil {
-			var v1res *runtime.VersionResponse
-			v1res, err = in.c.Version(ctrdutil.WithNamespace(ctx), &v1r)
-			if v1res != nil {
-				p, perr := v1res.Marshal()
-				if perr == nil {
-					resp := &runtime_alpha.VersionResponse{}
-					if perr = resp.Unmarshal(p); perr == nil {
-						res = resp
-					}
-				}
-				// actual error has precidence on error returned vs parse error issues
-				if perr != nil {
-					if err == nil {
-						err = perr
-					} else {
-						// extra log entry if convert response parse error and request error
-						log.G(ctx).WithError(perr).Error("Version failed")
-					}
-				}
-			}
-		}
+	var v1r runtime.VersionRequest
+	if err := alphaReqToV1Req(r, &v1r); err != nil {
+		return nil, errdefs.ToGRPC(err)
 	}
-	return res, errdefs.ToGRPC(err)
+
+	var v1res *runtime.VersionResponse
+	v1res, err = in.c.Version(ctrdutil.WithNamespace(ctx), &v1r)
+	if err != nil {
+		return nil, errdefs.ToGRPC(err)
+	}
+
+	resp := &runtime_alpha.VersionResponse{}
+	if err := v1RespToAlphaResp(v1res, resp); err != nil {
+		return nil, errdefs.ToGRPC(err)
+	}
+	return resp, nil
 }

 func (in *instrumentedService) UpdateRuntimeConfig(ctx context.Context, r *runtime.UpdateRuntimeConfigRequest) (res *runtime.UpdateRuntimeConfigResponse, err error) {
```

~~~~ :)

Reply:
I like it!

Reply:
just need to do the error precedence thing while still at least trying to do the conversion.. but yeah I like that to cut it down some!